docs/tutorials/smartpy-fa2-fungible/adding-metadata.md

```diff
@@ -0,0 +1,145 @@
+
+The completed contract that you create in this part is at [part_3_complete.py](https://github.com/trilitech/tutorial-applications/blob/main/smartpy_fa2_fungible/part_3_complete.py).
+
+## Getting a Pinata API key
+
+SmartPy test scenarios can upload files to IPFS with [Pinata](https://www.pinata.cloud/), so to follow this part of the tutorial, you need a free account on https://www.pinata.cloud and an API key.
+
+Follow these steps to create a Pinata API key:
+
+1. Create a free Pinata account at https://app.pinata.cloud/developers/api-keys.
+
+1. Go to the API Keys tab and click **New Key**.
+
+1. On the Create New API Key page, expand API Endpoint Access and enable the `pinFileToIPFS` permission, as in this picture:
+
+   <img src="/img/tutorials/pinata-key-permissions.png" alt="Selecting the permissions for the Pinata key" style={{width: 300}} />
+
+1. In the **Key Name** field, give the key a name, such as "My Key."
+
+1. Click **Create Key**.
+
+   The API Key Info window shows the API key and secret, which you must copy immediately, because it is not shown again.
+
+1. Copy the API Key and API Secret fields and save the values on your computer.
+You need these values in the next section.
+
+   You can see the new API key on the API Keys tab:
+
+   ![The new Pinata API key in the Pinata web app](/img/tutorials/created-pinata-key.png)
+
+## Uploading metadata
+
+The SmartPy `sp.pin_on_ipfs` function uploads metadata to IPFS via your Pinata account and "pins" it to ensure that it remains available until you remove it.
+In this section, you add code to use this function to upload and pin contract metadata to IPFS when the test scenario runs.
+It is still advisable to save a local copy of your metadata, just as you would with any upload service.
+
+1. After the code that instantiates the contract in the test scenario, use the `sp.create_tzip16_metadata` function to create a metadata object that is compatible with the [TZIP-16](https://gitlab.com/tezos/tzip/-/blob/master/proposals/tzip-16/tzip-16.md) standard for contract metadata:
+
+   ```smartpy
+   # Build contract metadata content
+   contract_metadata = sp.create_tzip16_metadata(
+       name="My FA2 fungible token contract",
+       description="This is an FA2 fungible token contract using SmartPy.",
+       version="1.0.0",
+       license_name="CC-BY-SA",
+       license_details="Creative Commons Attribution Share Alike license 4.0 https://creativecommons.org/licenses/by/4.0/",
+       interfaces=["TZIP-012", "TZIP-016"],
+       authors=["SmartPy <https://smartpy.io/#contact>"],
+       homepage="https://smartpy.io/ide?template=fa2_lib_fungible.py",
+       # Optionally, upload the source code to IPFS and add the URI here
+       source_uri=None,
+       offchain_views=contract.get_offchain_views(),
+   )
+   ```
+
+   This code must come after you create the contract because it uses the `contract.get_offchain_views` function to retrieve the contract's off-chain views.
```

Comment:
but before you add the contract to the scenario

Reply:
Changed this to "This code must come after you create the contract but before you add the contract to the scenario because"